Mathlib/CategoryTheory/CatCommSq.lean

```diff
@@ -0,0 +1,110 @@
+/-
+Copyright (c) 2023 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Equivalence
+
+/-!
+# 2-commutative squares of functors
+
+Similarly as `CommSq.lean` defines the notion of commutative squares,
+this file introduces the notion of 2-commutative squares of functors.
+
+If `T : C₁ ⥤ C₂`, `L : C₁ ⥤ C₃`, `R : C₂ ⥤ C₄`, `B : C₃ ⥤ C₄` are functors,
+then `[CatCommSq T L R B]` contains the datum of an isomorphism `T ⋙ R ≅ L ⋙ B`.
+
+Future work: using this notion in the development of the localization of categories
+(e.g. localization of adjunctions).
+
+-/
+
+namespace CategoryTheory
+
+open Category
+
+variable {C₁ C₂ C₃ C₄ C₅ C₆ : Type _} [Category C₁] [Category C₂] [Category C₃] [Category C₄]
+  [Category C₅] [Category C₆]
+  (T : C₁ ⥤ C₂) (L : C₁ ⥤ C₃) (R : C₂ ⥤ C₄) (B : C₃ ⥤ C₄)
+
+/-- `CatCommSq T L R B` expresses that there is a 2-commutative square of functors, where
+the functors `T`, `L`, `R` and `B` are respectively the left, top, right and bottom functors
+of the square. -/
+@[ext]
+class CatCommSq where
```

Comment:
I'm not completely sold on the idea that `CatCommSq` is a class. We have many classes where it's only _sometimes_ convenient to have it as a class, and I think we're not really happy with most of them. In the case of `Exact`, we demoted it to a structure, which was an improvement. In the case of `Epi`/`Mono`, I think the general consensus is that if we redid things, we wouldn't make them classes again, but changing it would be a huge refactor. Given that between `(h|v)Comp` and `(h|v)Inv` usage as a TC argument and an explicit argument is already inconsistent, my impulse would be to demote `CatCommSq` to a structure. For what it's worth, `CommSq` is also just a structure.

What do you think?

Reply:
For `hInv`, I made it an explicit argument because functors are usually not syntactically the `functor/inverse` field of an equivalence of categories, which would make quite inconvenient the use of a TC argument in that case.
For the applications, it is much more convenient to use `CatCommSq` as a TC. Otherwise, the number of parameters we would have to feed to later constructions would become unpractical. In the localization API, I already introduced the TC `Localization.Lifting` (see `CategoryTheory.Localization.Predicate`) which consists of some shapes of 2-commutative triangles of categories/functors. For consistency, I think 2-commutative squares should also be a TC.
(See also #6235 and #6236 for dependent PR which use `CatCommSq`. In later PRs, there will be a few `CatCommSq` instances also.)

Reply:
Sounds reasonable. Perhaps `hInv` should be stated in terms of `IsEquivalence`?

Reply:
Ok. Yes, I will try to adapt it with `[IsEquivalence]` and also provide the bijection `hInvEquiv` in a suitable form.

Reply:
I have slightly improved the code (avoiding some duplication), and added an instance in case we have `[IsEquivalence]`.